src/parallel.rs

```diff
@@ -133,33 +134,90 @@ impl TmpNodesReader {
 }
 
 impl ConcurrentNodeIds {
     /// Creates the ID generator starting at the given number.
     pub fn new(used: RoaringBitmap) -> ConcurrentNodeIds {
         let last_id = used.iter().last().map(|id| id + 1).unwrap_or(0);
-        let used = used.len();
+        let used_ids = used.len();
+        let available = RoaringBitmap::from_sorted_iter(0..last_id).unwrap() - used;
 
-        ConcurrentNodeIds { current: AtomicU32::new(last_id), used: AtomicU64::new(used) }
+        ConcurrentNodeIds {
+            current: Arc::new(AtomicU32::new(last_id)),
+            used: Arc::new(AtomicU64::new(used_ids)),
+            available,
+        }
     }
 
     /// Returns and increment the ID you can use as a NodeId.
-    pub fn next(&self) -> Result<u32> {
+    pub fn next(&mut self) -> Result<u32> {
         if self.used.fetch_add(1, Ordering::Relaxed) > u32::MAX as u64 {
             Err(Error::DatabaseFull)
         } else {
-            Ok(self.current.fetch_add(1, Ordering::Relaxed))
+            match self.available.select(0) {
+                Some(id) => {
+                    // remove_smallest has a faster access pattern than remove
+                    self.available.remove_smallest(1);
+                    Ok(id)
+                }
+                None => Ok(self.current.fetch_add(1, Ordering::Relaxed)),
+            }
         }
     }
 
     /// Returns the number of used ids in total.
     pub fn used(&self) -> u64 {
         self.used.load(Ordering::Relaxed)
     }
+
+    /// Split a concurrent node ids into `n` parts..
+    pub fn split_in(self, n: usize) -> Vec<ConcurrentNodeIds> {
+        if n <= 1 {
+            return vec![self];
+        }
+
+        let Self { current, used, available } = self;
+        let chunk_size = available.len() / n as u64;
+        let mut ret = Vec::new();
+
+        let mut iter = available.into_iter();
+        for _ in 0..(n - 1) {
+            let available = iter.by_ref().take(chunk_size as usize).collect();
+            ret.push(Self { current: current.clone(), used: used.clone(), available });
+        }
+        // the last element is going to contain everything remaining
+        ret.push(Self { current, used, available: iter.collect() });
+
+        ret
+    }
+
+    /// Recompose a vector of concurrent node ids into a single concurrent node ids.
+    /// Return `None` if called on an empty list.
+    pub fn merge(concurrents_node_ids: Vec<ConcurrentNodeIds>) -> Option<Self> {
+        concurrents_node_ids.into_iter().reduce(|left, right| Self {
+            current: left.current,
+            used: left.used,
+            available: left.available | right.available,
+        })
+    }
+}
+
+/// Cloning this type doesn't clone the available node ids
+impl Clone for ConcurrentNodeIds {
+    fn clone(&self) -> Self {
+        Self {
+            current: self.current.clone(),
+            used: self.used.clone(),
+            available: RoaringBitmap::new(),
+        }
+    }
```

Comment:
Why? We shouldn't implement the `Clone` trait but rather a special cloning method.

Reply:
If I implement a special cloning method then I can't call `repeatn` later on :cry: